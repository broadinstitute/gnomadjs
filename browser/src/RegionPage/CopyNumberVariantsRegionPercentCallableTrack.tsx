import React from 'react'

import { referenceGenome } from '@gnomad/dataset-metadata/metadata'
import Query from '../Query'
import DiscreteBarPlot from '../DiscreteBarPlot'

type OwnProps = {
  datasetId: string
  chrom: string
  start: number
  stop: number
}

// @ts-expect-error TS(2456) FIXME: Type alias 'Props' circularly references itself.
type Props = OwnProps & typeof CopyNumberVariantsRegionPercentCallableTrack.defaultProps

// @ts-expect-error TS(7022) FIXME: 'CopyNumberVariantsRegionPercentCallableTrack' implicitly has type '... Remove this comment to see the full error message
const CopyNumberVariantsRegionPercentCallableTrack = ({ datasetId, chrom, start, stop }: Props) => {
  const operationName = 'CopyNumberVariantsRegionPercentCallableTrack'
  const query = `
  query ${operationName}($datasetId: CopyNumberVariantDatasetId!, $chrom: String!, $start: Int!, $stop: Int!, $referenceGenome: ReferenceGenomeId!) {
    region(chrom: $chrom, start: $start, stop: $stop, reference_genome: $referenceGenome) {
            cnv_track_callable_coverage(dataset: $datasetId) {
                xpos
                percent_callable
                position
                contig
            }
        }
    }
`
  return (
    <Query
      operationName={operationName}
      query={query}
      variables={{ datasetId, chrom, start, stop, referenceGenome: referenceGenome(datasetId) }}
      loadingMessage="Loading coverage"
      loadingPlaceholderHeight={220}
      errorMessage="Unable to load coverage"
      success={(data: any) => {
        return data.region && data.region.cnv_track_callable_coverage
      }}
    >
      {({ data }: any) => {
        const transformedArray = data.region.cnv_track_callable_coverage.map((item: any) => ({
          pos: item.position,
          percent_callable: item.percent_callable,
          xpos: item.xpos,
        }))
        transformedArray.sort((a: any, b: any) => a.xpos - b.xpos)

        const coverage = [
          {
            color: 'rgb(70, 130, 180)',
            buckets: transformedArray,
            name: 'percent callable',
            opacity: 0.7,
          },
        ]

        return (
          <DiscreteBarPlot
            datasets={coverage}
            height={200}
            regionStart={start}
            regionStop={stop}
            chrom={Number(chrom)}
          />
        )
      }}
    </Query>
  )
}

export default CopyNumberVariantsRegionPercentCallableTrack
