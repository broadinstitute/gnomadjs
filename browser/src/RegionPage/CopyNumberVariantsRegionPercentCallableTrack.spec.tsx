import React from 'react'
import { createRenderer } from 'react-test-renderer/shallow'

import { jest, describe, expect, test } from '@jest/globals'
import { mockQueries } from '../../../tests/__helpers__/queries'
import Query, { BaseQuery } from '../Query'

import CopyNumberVariantsRegionPercentCallableTrack from './CopyNumberVariantsRegionPercentCallableTrack'

import { allDatasetIds, hasCopyNumberVariantCoverage } from '@gnomad/dataset-metadata/metadata'

jest.mock('../Query', () => {
  const originalModule = jest.requireActual('../Query')

  return {
    __esModule: true,
    ...(originalModule as object),
    default: jest.fn(),
    BaseQuery: jest.fn(),
  }
})

const { resetMockApiCalls, resetMockApiResponses, simulateApiResponse, setMockApiResponses } =
  mockQueries()

beforeEach(() => {
  Query.mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('Query', query, children, operationName, variables)
    )
  )
  ;(BaseQuery as any).mockImplementation(
    jest.fn(({ query, children, operationName, variables }) =>
      simulateApiResponse('BaseQuery', query, children, operationName, variables)
    )
  )
})

afterEach(() => {
  resetMockApiCalls()
  resetMockApiResponses()
})

const datasetsWithCoverage = allDatasetIds.filter((datasetId) =>
  hasCopyNumberVariantCoverage(datasetId)
)

describe.each(datasetsWithCoverage)(
  'CopyNumberVariantsRegionPercentCallableTrack with dataset %s',
  (datasetId) => {
    test('queries with appropriate params', () => {
      setMockApiResponses({
        CopyNumberVariantsRegionPercentCallableTrack: () => ({
          region: {
            cnv_track_callable_coverage: [],
          },
        }),
      })

      const shallowRenderer = createRenderer()
      shallowRenderer.render(
        <CopyNumberVariantsRegionPercentCallableTrack
          datasetId={datasetId}
          chrom={1}
          start={123}
          stop={456}
        />
      )

      expect(shallowRenderer.getRenderOutput()).toMatchSnapshot()
    })
  }
)
