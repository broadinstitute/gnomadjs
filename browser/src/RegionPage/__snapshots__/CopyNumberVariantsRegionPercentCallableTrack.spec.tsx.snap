// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CopyNumberVariantsRegionPercentCallableTrack with dataset gnomad_cnv_r4 queries with appropriate params 1`] = `
<mockConstructor
  errorMessage="Unable to load coverage"
  loadingMessage="Loading coverage"
  loadingPlaceholderHeight={220}
  operationName="CopyNumberVariantsRegionPercentCallableTrack"
  query="
  query CopyNumberVariantsRegionPercentCallableTrack($datasetId: CopyNumberVariantDatasetId!, $chrom: String!, $start: Int!, $stop: Int!, $referenceGenome: ReferenceGenomeId!) {
    region(chrom: $chrom, start: $start, stop: $stop, reference_genome: $referenceGenome) {
            cnv_track_callable_coverage(dataset: $datasetId) {
                xpos
                percent_callable
                position
                contig
            }
        }
    }
"
  success={[Function]}
  variables={
    {
      "chrom": 1,
      "datasetId": "gnomad_cnv_r4",
      "referenceGenome": "GRCh38",
      "start": 123,
      "stop": 456,
    }
  }
>
  [Function]
</mockConstructor>
`;
